content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+Airbus, an aerospace corporation, uses CompCert to write the software for its planes. A [2014 report][airbus] from the company indicated that the worst-case execution time of its programs had improved by 12%. The commercial success of CompCert speaks to its ability to hold its own as an optimizing compiler, although it still doesn't employ some more aggressive code optimizations. 
+
+### An Achilles' Heel: Separate Compilation
+
+Consider the following two programs.
+
+`a.c`:
+
+    #include <stdio.h>
+    int x;
+    extern int* const xptr;
+    int main() {
+        x = 1;
+        *xptr = 0;
+        printf("%d\n", x + x);
+        return 0;
+    }
+
+`b.c`:
+
+    #include <stdio.h>
+    extern int x;
+    int* const xptr = &x;
+
+When CompCert (version 2.4) compiles these programs separately and they are linked, the output is `2`, when it should actually be `0`. The global variable `xptr` is a `const` in `a.c`, so CompCert's value analysis assumes it's always uninitialized and assigns it $\bot$. Therefore, `xptr` is deemed unaliasable with `x`, and in `a.c`, `x` is presumed to be `1`. Constant propagation then replaces `x + x` with `2`. `xptr`'s declaration as `extern` means that `b.c` is able to initialize it, aliasing `x` and `xptr`. This bug, discovered by the authors of [SepCompCert][scc] in their efforts to verify separate compilation, is astoundingly egregious. It raises the question: how do we know that compiler correctness theorems encompass the behavior exhibited by the compiler? Relatedly, when do we know if a language specification characterizes its programs correctly? 
+
+I don't know the answers to those questions, or the ones I'm asking you for discussion. I don't think that writing compilers in Coq is going to become the standard, but I do think that developing a formal semantics for several intermediate languages will lower the threshold for verifying compilers. CompCert has changed the way we approach compiler correctness and proof engineering for the better, and I can't wait to see where the next decade takes us in this field. 
+
+## Questions
+
+* Xavier Leroy undertook the feat of engineering CompCert as a research endeavor. Should he have spent that time differently? Should implementation be left to industry? Does theoretical ideation hold value without the promise of practical use? 
+* In some sense, verifying software becomes a "turtles all the way down" problem where different tools must be assumed correct (or in need of verification). For instance, were a bug to be found in Coq, the correctness of CompCert would be meaningless. Does this assessment of the work resonate with you? Do you think it's unfair to characterize compiler correctness efforts this way? What makes CompCert valuable if we don't assume that the tools used are correct, if anything?
```

Comment:
The key word in discussions like this is typically ["TCB"](https://en.wikipedia.org/wiki/Trusted_computing_base) as in "how small is small enough for a compiler's TCB?"